workers/unity/Packages/com.improbable.gdk.core/Worker/Worker.cs

```diff
@@ -10,77 +10,137 @@ namespace Improbable.Gdk.Core
 
-        public static Worker Connect(ConnectionConfig config, ILogDispatcher logDispatcher, Vector3 origin)
-        {
-            return new Worker(config, logDispatcher, origin);
+                var deploymentName = deploymentListCallback(deploymentList);
+                var connectionParams = config.CreateConnectionParameters();
+                using (var connectionFuture = locator.ConnectAsync(deploymentName, connectionParams, queueCallback))
+                {
+                    var connection = await Task.Run(() => connectionFuture.Get());
+                    if (!connection.IsConnected)
+                    {
+                        throw new ConnectionFailedException(GetConnectionFailureReason(connection),
+                            ConnectionErrorReason.CannotEstablishConnection);
+                    }
+
+                    var worker = new Worker(config.WorkerType, connection, logger, origin);
+                    worker.AddCoreSystems();
+                    return worker;
+                }
+            }
         }
 
-        public static void Disconnect(Worker worker)
+        public bool TryGetEntity(EntityId entityId, out Entity entity)
         {
-            worker.Dispose();
+            return World.GetExistingManager<WorkerSystem>().TryGetEntity(entityId, out entity);
         }
 
         public void Dispose()
         {
-            EntityMapping.Clear();
             World?.Dispose();
             World = null;
-            OnDisconnect?.Invoke(this);
-            ConnectionUtility.Disconnect(Connection);
             Connection?.Dispose();
             Connection = null;
         }
+
+        private static async Task<DeploymentList> GetDeploymentList(Locator locator)
+        {
+            using (var deploymentsFuture = locator.GetDeploymentListAsync())
+            {
+                var deploymentList = await Task.Run(() => deploymentsFuture.Get());
+                if (!string.IsNullOrEmpty(deploymentList.Error))
+                {
+                    throw new ConnectionFailedException(deploymentList.Error, ConnectionErrorReason.DeploymentNotFound);
+                }
+
+                return deploymentList;
+            }
+        }
+
+        private static string GetConnectionFailureReason(Connection connection)
+        {
+            var connectionFailureReason = "No reason found for connection failure";
+            var dispatcher = new Dispatcher();
+            dispatcher.OnDisconnect(op => connectionFailureReason = op.Reason);
+            dispatcher.Process(connection.GetOpList(0));
+
+            return connectionFailureReason;
+        }
+
+        private void AddCoreSystems()
+        {
+            var workerSystem = World.GetOrCreateManager<WorkerSystem>();
+            workerSystem.Connection = Connection;
+            workerSystem.LogDispatcher = LogDispatcher;
+            workerSystem.Origin = Origin;
+            workerSystem.WorkerType = WorkerType;
```

Comment:
we are now duplicating fields between two classes and the user can modify them in the system. This could cause quite a few problems when the user suddenly modifies one of them..

Reply:
I think this is a fridge error. We can make them readonly to prevent that, but if a user goes out of their way to break things then things will break. Things are duplicated because one is a wrapper around the other. We can make it so that the Worker class has properties to call these on each access if we need them to be the same but it makes cleanup slightly different.

Reply:
Changed it to use a constructor so it can be readonly.